aptos-move/framework/aptos-stdlib/sources/cryptography/groups.move

```diff
@@ -0,0 +1,744 @@
+    use std::option::Option;
+
+    // Error codes
+    const E_NATIVE_FUN_NOT_AVAILABLE: u64 = 1;
+    const E_UNKNOWN_GROUP: u64 = 2;
+    const E_UNKNOWN_PAIRING: u64 = 3;
+
+    /// `BLS12_381_G1` represents a group used in BLS12-381 pairing.
+    /// The group is a prime-order group on an elliptic curve `y^2=x^3+4` defined over `Fq`.
+    /// `Fq` is a finite field with `q=0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab`.
+    /// THe order of the group `r` is 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001.
+    /// There exists a bilinear mapping from `(BLS12_381_G1, BLS12_381_G2)` to `BLS12_381_Gt`.
+    ///
+    /// A `Scalar<BLS12_381_G1>` is an integer between 0 and `r-1`.
+    ///
+    /// Function `deserialize_scalar<BLS12_381_G1>` and `serialize_scalar<BLS12_381_G1>`
+    /// assumes a 32-byte little-endian encoding of a `Scalar<BLS12_381_G1>`.
+    ///
+    /// An `Element<BLS12_381_G1>` is an element in `G1`.
+    ///
+    /// Function `serialize_element_uncompressed<BLS12_381_G1>` and `deserialize_element_uncompressed<BLS12_381_G1>`
+    /// assumes a 96-byte encoding `[b_0, ..., b_95]` of an `Element<BLS12_381_G1>`, with the following rules.
+    /// - `b_95 & 0x40` is the infinity flag.
+    /// - The infinity flag is 1 if and only if the element is the point at infinity.
+    /// - The infinity flag is 0 if and only if the element is a point `(x,y)` on curve `E(Fq)`, with the following rules.
+    ///     - `[b_0, ..., b_47 & 0x3f]` is a 48-byte little-endian encoding of `x`.
+    ///     - `[b_48, ..., b_95 & 0x3f]` is a 48-byte little-endian encoding of 'y'.
+    ///
+    /// Function `serialize_element_compressed<BLS12_381_G1>` and `deserialize_element_compressed<BLS12_381_G1>`
+    /// assumes a 48-byte encoding `[b_0, ..., b_47]` of an `Element<BLS12_381_G1>` with the following rules.
+    /// - `b_47 & 0x40` is the infinity flag.
+    /// - The infinity flag is 1 if and only if the element is the point at infinity.
+    /// - The infinity flag is 0 if and only if the element is a point `(x,y)` on curve, with the following rules.
+    ///     - `[b_0, ..., b_47 & 0x3f]` is a 48-byte little-endian encoding of `x`.
+    ///     - `b_47 & 0x80` is the positiveness flag.
+    ///     - The positiveness flag is 1 if and only if `y > -y`.
+    struct BLS12_381_G1 {}
+
+    /// `BLS12_381_G2` represents a group used in BLS12-381 pairing.
+    /// The group is a prime-order group on an elliptic curve `y^2=x^3+4(u+1)` defined over `Fq2`.
+    /// `Fq2` is an extension field of `Fq`, constructed as `Fq2=Fq[u]/(u^2+1)`.
+    /// `Fq` is a finite field with `q=0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab`.
+    /// THe order of the group `r` is 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001.
+    /// There exists a bilinear mapping from `(BLS12_381_G1, BLS12_381_G2)` to `BLS12_381_Gt`.
```

Comment:
You are repeating this. I would remove it.

Reply:
If we compress the shared bls12-381 preliminaries (into ~5 lines), we can keep them inline with G1/G2/Gt structs, and save a few scrolling/clicks for the readers?

If we pull out the shared part, I actually don't know where they should go.

Reply:
I'll leave it up to you. It's just slightly repetitive to me. Perhaps, start the sentence with _"As mentioned before, there exists a bilinear map from [...]"_